src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestStage.java

```diff
@@ -272,6 +280,53 @@ synchronized SnowflakeFileTransferMetadataWithAge refreshSnowflakeMetadata(boole
     }
   }
 
+  /**
+   * GCS requires a signed url per file. We need to fetch this from the server for each put
+   *
+   * @throws SnowflakeSQLException
+   * @throws IOException
+   */
+  SnowflakeFileTransferMetadataV1 fetchSignedURL(String fileName)
```

Comment:
Thanks for doing this! I feel like the logic could be combined further, basically the idea is that we will need a force refresh for uploading every file, so instead of adding `fetchSignedURL`, why can't you just call `refreshSnowflakeMetadata` before `uploadWithoutConnection` for every file? Please give it a thought and see if it makes sense